### ISSUE TITLE
Allow using binding.bp to start a debug session

### ISSUE THREAD

**Issue by st0012** [ruby/debug#65]
As stated in the readme, the correct way to start a debug session in the middle of a program is:

```ruby
require "debug"
DEBUGGER__.console
```

And then the user can freely use `binding.bp` to insert new breakpoints.

This is a fine workflow, but I think we can make it more simpler. Since we already have the `binding.bp` method, which just looks like `byebug` and `binding.pry`, why don't we make it work like them?

```ruby
require "pry"
binding.pry

require "byebug"
byebug

require "debug"
binding.bp
```

Then the users will have one less thing to learn 🙂 

(I have to admit that it took me a while to remember to call `DEBUGGER__.console`, which I still forget from time to time 😂)

**Comment by ko1** [comment]
Yes, it is planned idea.

The drawback is, explicit debug start will track the program execution:

* Thread creation
* Code loading including `eval` (store loaded Ruby code to show)
* And more (not yet)

In many cases it is not a problem.

-----
Terminology:

* Load debug functionality: Loading the debug classes/modules/methods
  * `require 'debug'`
  * `require 'debug/session'`
* Start debug: load debug functionality and start tracking the information for debugging. Also load `.rdbgrc` for configuration.
  * `require 'debug/run'`
  * `DEBUGGER__.console`
* Stop in a debugger context:
  * breakpoint
  * `binding.bp`
  * Ctrl-C on (remote) console

----

Maybe this proposal is, if `binding.bp` is called without start debugging, start debugging and stop at this location.

**Comment by ko1** [comment]
Another possibility is, `require 'debug'` start debgging but don't stop the program (== `RUBY_DEBUG_NONSTOP` mode).
The drawback is, if user doesn't want to collect information (don't want to consume any overhead even if `require 'debug' is called), the interpreter starts by adding `gem 'debug'` in `Gemfile`.

**Comment by ko1** [comment]
In a future, like `binding.irb`, I want to introduce it as standard. But there is an issue describe above.

**Comment by st0012** [comment]
> Maybe this proposal is, if binding.bp is called without start debugging, start debugging and stop at this location.

That's what I have in mind too. Should I open a PR for it?

**Comment by ko1** [comment]
I showed some points, and we need to choose one.

**Comment by st0012** [comment]
Ah sorry I didn't stress that clearly: I took that line as 

`if binding.bp is called without start debugging, start "debugging session" and stop at this location.`, which is like

```ruby
# move this out from Session.initialize_session
Binding.module_eval do
  def bp command: nil
    session_initialized = ::DEBUGGER__.const_get(:SESSION)

    ::DEBUGGER__.console unless session_initialized

    if command
      cmds = command.split(";;")
      SESSION.add_initial_commands cmds
    end

    ::DEBUGGER__.add_line_breakpoint __FILE__, __LINE__ + 1, oneshot: true if session_initialized
    true
  end
end
```

**Comment by ko1** [comment]
Yes. I know. And I showed some drawback https://github.com/ruby/debug/issues/65#issuecomment-855156514

**Comment by st0012** [comment]
let me summarize our options here:

1. start the debugging session within `binding.bp` call if it's not started yet.

> if binding.bp is called without start debugging, start "debugging session" and stop at this location.

2. start the debugging session when this gem is required.
 
> require 'debug' start debgging but don't stop the program (== RUBY_DEBUG_NONSTOP mode).

and the difference between then is the timing of debugging session starts. the 1st options starts at the first `binding.bp` call and the 2nd one starts whenever this lib is required (which is the boot time in Rails app). 

I think the 1st option makes more sense and I'd do it like I mentioned in https://github.com/ruby/debug/issues/65#issuecomment-855330594

**Comment by ko1** [comment]
> I think the 1st option makes more sense

Could you explain why?

**Comment by st0012** [comment]
activating the debug session with `require "debug"` means that Rails users (or anyone uses `Bundler.require`) have `gem "debug"` in their Gemfiles could activate the session without knowing it.

I know asking them to add `, require: false` can workaround this issue, but


```ruby
gem "debug", require: false

# wherever I want to debug
require "debug" # start the session
binding.bp # and add a breakpoint
```

still has more code than

```ruby
gem "debug" # require files & define binding.bp, but don't start anything

# wherever I want to debug
binding.bp # start the session and add a breakpoint
```

**Comment by ko1** [comment]
You are right. How about disadvantages?

**Comment by st0012** [comment]
I don't see a clear downside in this approach. shouldn't

```ruby
gem "debug"

binding.bp
```

essentially the same as this?

```ruby
require "debug"

DEBUGGER__.console

# and later
binding.bp
```

the only difference seems to be that we'll patch `Binding` right after `require "debug"` instead of `DEBUGGER__.console`.

can you point out anything I miss here?

**Comment by ko1** [comment]
https://github.com/ruby/debug/issues/65#issuecomment-855156514

> Thread creation
> Code loading including eval (store loaded Ruby code to show)
> And more (not yet)

**Comment by st0012** [comment]
I don't understand. since these start in `Session#initialize`, doesn't `DEBUGGER__.console` starts them as well?

**Comment by ko1** [comment]
In https://github.com/ruby/debug/issues/65#issuecomment-855156514 I showed 3 points.

1. Load debug functionality: Loading the debug classes/modules/methods
2. Start debug: load debug functionality and start tracking the information for debugging. Also load .rdbgrc for configuration.
3. Stop in a debugger context:

* requiring `debig` start (1)
* requiring `debug/run` (or `exe/rdbg`, `console` and so on) start (1) and (2)
* `binding.bp` without starting (2), start (2) and (3).

So *after* `binding.bp`, information is tracked. However, the code running *before* first `binding.bp` without (2) are not collected. And as I wrote,

> In many cases it is not a problem.

In other words, there is a problem in a few cases.

**Comment by st0012** [comment]
because `DEBUGGER__.console` also stops the program, so the current flow looks like this, right?

```ruby
require "debug" # code loaded (1)

# nothing tracked here

DEBUGGER__.console # start tracking (2) + stop the program (3)

# things are tracked

binding.bp # stop the program (3)
```

to users, it means that "before the program is first stopped, nothing is being tracked". 

so if we can make `binding.bp` do both (2) and (3) properly, it's still the same behavior users

```ruby
require "debug" # code loaded (1)

# nothing tracked here

binding.bp # start tracking (2) + stop the program (3)

# things are tracked

binding.bp # stop the program (3)
```

**Comment by ko1** [comment]
or just `.start` I'm thinking.

> because DEBUGGER__.console also stops the program, so the current flow looks like this, right?

This is why there is `-n` option on rdbg.

**Comment by st0012** [comment]
> This is why there is -n option on rdbg.

but it's not accessible from application commands like `rails` or `rspec`. I think in the future most of the web developer users would not start the debugging session with the `rdbg`, so `-n` won't be an option to them.

**Comment by ko1** [comment]
not important but they are also `-n`:

* `nonstop` option for console and so on.
* `RUBY_DEBUG_NONSTOP` envval.

I'm thinking to start debug session (tracking) only with `require 'debug'`.

**Comment by ko1** [comment]
* 1. start tracking
  * `require 'debug'`
  * `gem 'debug'` in `Gemfile` with `Bundler.require`
* 2. start console
  * a. `require 'debug/run'` or `require 'debug/open'`
  * b. `console` for local console
  * c. `open` methods for remote
  * d. `binding.bp`

**Comment by st0012** [comment]
> * 1. start tracking
>   
>   
>   * `require 'debug'`
>   * `gem 'debug'` in `Gemfile` with `Bundler.require`
> * 1. start console
>   
>   
>   * a. `require 'debug/run'` or `require 'debug/open'`
>   * b. `console` for local console
>   * c. `open` methods for remote
>   * d. `binding.bp`

yeah I think this will make starting the debugger a lot easier 👍 
since this may conflict with the DAP work, I think it's better to be implemented by you as well?

**Comment by st0012** [comment]
@ko1 I'm going to give it a shot 🙂

**Comment by st0012** [comment]
@ko1 can you help me verify if these scenarios are correct?

## Scenarios

### `-r debug/run`

Script:

```ruby
a = 1 # stops here
b = 2
c = 3
d = 4
```

Command: `$ ruby -Ilib -r debug/run target.rb`

### `-r debug/open`

Script:

```ruby
a = 1 # stops here
b = 2
c = 3
d = 4
```

Commands: 
- `$ ruby -Ilib -r debug/open target.rb`
- `$ exe/rdbg --attach`

### `require 'debug/run'`

Script:

```ruby
a = 1
b = 2
require "debug/run" # stops here
c = 3
d = 4
```

Command: `$ ruby -Ilib target.rb`

### `require 'debug/open'`

Script:

```ruby
a = 1
b = 2
require "debug/open" # stops here
c = 3
d = 4
```

Commands:
- `$ ruby -Ilib target.rb`
- `$ exe/rdbg --attach`

### `DEBUGGER__.console`

Script:

```ruby
require "debug"
a = 1
b = 2
DEBUGGER__.console # stops here
c = 3
d = 4
```

Command: `$ ruby -Ilib target.rb`

### `DEBUGGER__.open`

Script:

```ruby
require "debug"
a = 1
b = 2
DEBUGGER__.open # stops here
c = 3
d = 4
```

Commands: 
- `$ ruby -Ilib target.rb`
- `$ exe/rdbg --attach`


### `binding.bp`

Script:

```ruby
require "debug"
a = 1
b = 2
binding.bp # stops here
c = 3
d = 4
```

Command: `$ ruby -Ilib target.rb`

**Comment by ko1** [comment]
`.console` and `.open` should be at the beginning of the program, I assumed.

**Comment by st0012** [comment]
does that mean in the future those 2 methods won't stop the program anymore? (which I'd prefer)

**Comment by ko1** [comment]
I also find that entering debug console with `C-c` should be affect with `require 'debug'`. Now I'm making it to allow entering the console.

**Comment by ko1** [comment]
I found that the application should terminate with SIGINT if user doesn't enable debugger explicitly.

* Implicit
  * `gem debug` in Gemfile
  * `require 'debug'`
* Explicit
  * use `rdbg` command
  * `require 'debug/run'`
  * `require 'debug/open'`

-----

Allowed operations:

* Implicit
  * stop on `binding.bp` like `binding.pry/irb`
* Explicit
  * stop on `C-c`
  * stop at the beggining
* Not sure
  * (1) start tracing (sources / threads / ...)
  * (2) load `~/.rdbginit`

**Comment by ko1** [comment]
Now I'm thinking to enable (1) and (2) even if it is implicit.

**Comment by ko1** [comment]
To control them, I'm thinking `DEBUGGER__::start()` method:

```ruby
# debug/session.rb

# start debugger functionality.
#
#  trace_loading: start tracing loading files.
#  trace_threading: start tracing thread creation/termination
#  load_init: allow to load init
#  suspend_on_sigint: enter debug session with C-c
#  nonstop: Do not stop at the beggining of main script
def self.start trace_loading: true, 
               trace_threading: true,
               load_init: true,
               suspend_on_sigint: false,
               nonstop: true
end

def self.run
  start suspend_on_sigint: true, nonstop: true
end
```

```ruby
# debug.rb
require 'debug/session'
DEBUGGER__::start

# debug/run.rb
require 'debug/session'
DEBUGGER__::run

# debug/open.rb
require 'debug/session'
DEBUGGER__::open
```

**Comment by ko1** [comment]
I'm not sure anybody wants to use `trace_loading` and `trace_threading`, so now I don't introduce it.

**Comment by st0012** [comment]
@ko1 thank you!

**Comment by ko1** [comment]
Please try it and tell me how it good or not.

**Comment by st0012** [comment]
@ko1 just tested in my Rails project and it worked great 👍
